spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/actions/BaseGetChangeSetSparkAction.java

```diff
@@ -0,0 +1,282 @@
+import org.apache.iceberg.relocated.com.google.common.collect.Sets;
+import org.apache.iceberg.spark.FileScanTaskSetManager;
+import org.apache.iceberg.spark.SparkReadOptions;
+import org.apache.iceberg.util.SnapshotUtil;
+import org.apache.spark.sql.Dataset;
+import org.apache.spark.sql.Row;
+import org.apache.spark.sql.SparkSession;
+import org.apache.spark.sql.functions;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import static org.apache.spark.sql.functions.lit;
+
+public class BaseGetChangeSetSparkAction extends BaseSparkAction<GetChangeSet, GetChangeSet.Result>
+    implements GetChangeSet {
+  private static final Logger LOG = LoggerFactory.getLogger(BaseGetChangeSetSparkAction.class);
+  public static final String RECORD_TYPE = "_record_type";
+  public static final String COMMIT_SNAPSHOT_ID = "_commit_snapshot_id";
+  public static final String COMMIT_TIMESTAMP = "_commit_timestamp";
+  public static final String COMMIT_ORDER = "_commit_order";
+
+  private final List<Long> snapshotIds = Lists.newLinkedList();
+  private final Table table;
+  private final List<Dataset<Row>> dfs = Lists.newLinkedList();
+
+  private boolean ignoreRowsDeletedWithinSnapshot = true;
+  private Expression filter = Expressions.alwaysTrue();
+
+  protected BaseGetChangeSetSparkAction(SparkSession spark, Table table) {
+    super(spark);
+    this.table = table;
+  }
+
+  @Override
+  public Result execute() {
+    for (int i = 0; i < snapshotIds.size(); i++) {
+      generateCdcRecordsPerSnapshot(snapshotIds.get(i), i);
+    }
+
+    Dataset<Row> outputDf = null;
+    for (Dataset<Row> df : dfs) {
+      if (outputDf == null) {
+        outputDf = df;
+      } else {
+        outputDf = outputDf.unionByName(df, true);
+      }
+    }
+    return new BaseGetChangeSetSparkActionResult(outputDf);
+  }
+
+  private void generateCdcRecordsPerSnapshot(long snapshotId, int commitOrder) {
+    Snapshot snapshot = table.snapshot(snapshotId);
+    if (snapshot.operation().equals(DataOperations.REPLACE)) {
+      return;
+    }
+
+    // metadata deleted data files
+    Dataset<Row> deletedDf = readMetadataDeletedFiles(snapshotId, commitOrder);
+    if (deletedDf != null) {
+      dfs.add(deletedDf);
+    }
+
+    // pos and eq deletes
+    Dataset<Row> rowLevelDeleteDf = readRowLevelDeletes(snapshotId, commitOrder);
+    if (rowLevelDeleteDf != null) {
+      dfs.add(rowLevelDeleteDf);
+    }
+
+    // new data file as the insert
+    Dataset<Row> df = readAppendDataFiles(snapshotId, commitOrder);
+    if (df != null) {
+      dfs.add(df);
+    }
+  }
+
+  private Dataset<Row> readAppendDataFiles(long snapshotId, int commitOrder) {
+    List<FileScanTask> fileScanTasks = planAppendedFiles(snapshotId);
+    if (fileScanTasks.isEmpty()) {
+      return null;
+    }
+
+    String groupID = UUID.randomUUID().toString();
+    FileScanTaskSetManager manager = FileScanTaskSetManager.get();
+    manager.stageTasks(table, groupID, fileScanTasks);
```

Comment:
We stage tasks but we never invalidate them. This will cause a memory leak. It will be tricky to invalidate them as we don't know whether it is already safe to do so. We can instruct the manager to remove tasks after the first access or make the action result closable. Let's think what will be best.